jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,343 @@
+         *
+         * @return the labels of all vertices of the path
+         */
+        public Set<Integer> getLabels() {
+            return labels;
+        }
+
+        /**
+         * Returns the cost of the path
+         *
+         * @return the cost of the path
+         */
+        public double getCost() {
+            return cost;
+        }
+
+        /**
+         * Returns a shallow copy of this labeled path instance. Vertices are not cloned.
+         *
+         * @return a shallow copy of this path.
+         *
+         * @throws RuntimeException in case the clone is not supported
+         *
+         * @see java.lang.Object#clone()
+         */
+        public LabeledPath<V> clone() {
+            try {
+                LabeledPath<V> newLabeledPath = TypeUtil.uncheckedCast(super.clone());
+                newLabeledPath.vertices = (ArrayList<V>) this.vertices.clone();
+                newLabeledPath.labels = new HashSet<>(this.labels);
+                newLabeledPath.cost = this.cost;
+
+                return newLabeledPath;
+            } catch (CloneNotSupportedException e) {
+                e.printStackTrace();
+                throw new RuntimeException();
+            }
+        }
+    }
+
+    private Graph<V, E> graph;
+    private Map<V, Integer> labels;
+    private int lengthBound;
+
+    /**
+     * Constructs an algorithm with given inputs
+     *
+     * @param graph the (improvement) graph on which to calculate the local augmentation
+     * @param lengthBound the upper bound for the length of cycles to detect
+     * @param labels the labels of the vertices encoding the subsets of vertices
+     */
+    public AhujaOrlinSharmaCyclicExchangeLocalAugmentation(Graph<V, E> graph, int lengthBound, Map<V, Integer> labels) {
+        this.graph = Objects.requireNonNull(graph, "Graph cannot be null");
+        this.lengthBound = lengthBound;
+        this.labels = Objects.requireNonNull(labels, "Labels cannot be null");
+    }
+
+    /**
+     * Calculates a valid subset-disjoint negative cycle.
+     * If there is no such cycle, it returns an empty LabeledPath instance with cost Double.MAX_VALUE.
+     *
+     * @return a valid subset-disjoint negative cycle encoded as LabeledPath
+     */
+    public GraphWalk<V, E> getLocalAugmentationCycle() {
+
+        int k = 1;
+        LabeledPath<V> cycleResult = new LabeledPath<>(lengthBound);
+
+        /*
+         * Store the path sets via a with Pair<V, V> as for the head and tail of the path such that the domination tests runs faster.
+         * Furthermore, use a TreeSet to sort for the cost of the paths such that the domination test stops optimally concerning the cost of the paths.
+         */
+        Map<Pair<V, V>, TreeSet<LabeledPath<V>>> pathsLengthK = new LinkedHashMap<>();
+        Map<Pair<V, V>, TreeSet<LabeledPath<V>>> pathsLengthKplus1 = new LinkedHashMap<>();
+
+        // initialize pathsLengthK for k = 1
+        for(E e : graph.edgeSet()) {
```

Comment:
Is the input graph allowed to have self-loops?  If so, then what happens if one of these edges already forms a cycle?  If not, then that should be a sanity check.